src/rust/src/rustzcash.rs

```diff
@@ -449,34 +455,38 @@ pub extern "C" fn librustzcash_sapling_compute_cmu(
     true
 }
 
-/// Computes \[sk\] \[8\] P for some 32-byte point P, and 32-byte Fs.
+/// Computes KDF^Sapling(KA^Agree(sk, P), ephemeral_key).
 ///
 /// If P or sk are invalid, returns false. Otherwise, the result is written to
 /// the 32-byte `result` buffer.
 #[no_mangle]
-pub extern "C" fn librustzcash_sapling_ka_agree(
+pub extern "C" fn librustzcash_sapling_ka_derive_symmetric_key(
     p: *const [c_uchar; 32],
     sk: *const [c_uchar; 32],
+    ephemeral_key: *const [c_uchar; 32],
     result: *mut [c_uchar; 32],
 ) -> bool {
-    // Deserialize p
-    let p = match de_ct(jubjub::ExtendedPoint::from_bytes(unsafe { &*p })) {
-        Some(p) => p,
+    // Deserialize p (either ephemeral_key or pk_d; we can handle them identically).
+    let epk = match SaplingDomain::<Network>::epk(&EphemeralKeyBytes(unsafe { *p })) {
```

Comment:
[SaplingDomain::epk](https://github.com/zcash/librustzcash/blob/zcash_primitives-0.10.0/zcash_primitives/src/sapling/note_encryption.rs#L241-L246) in zcash_primitives 0.10.0 calls `sapling::keys::EphemeralPublicKey::from_bytes(...).into()`. The `EphemeralPublicKey` wraps a `jubjub::ExtendedPoint`; its `from_bytes` is implemented [here](https://github.com/zcash/librustzcash/blob/zcash_primitives-0.10.0/zcash_primitives/src/sapling/keys.rs#L390-L392) and `.into()` just converts from a `CtOption` to an `Option`. So this is equivalent to the original code which was not doing any validation beyond checking for a Jubjub point.